plugin/trino-hive/src/test/java/io/trino/plugin/hive/crypto/TestHiveParquetEncryption.java

```diff
@@ -0,0 +1,168 @@
+import io.trino.testing.MaterializedResult;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.parquet.crypto.ColumnEncryptionProperties;
+import org.apache.parquet.crypto.FileEncryptionProperties;
+import org.apache.parquet.crypto.ParquetCipher;
+import org.apache.parquet.example.data.Group;
+import org.apache.parquet.example.data.simple.SimpleGroupFactory;
+import org.apache.parquet.hadoop.ParquetWriter;
+import org.apache.parquet.hadoop.example.ExampleParquetWriter;
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.schema.MessageTypeParser;
+import org.junit.jupiter.api.Test;
+
+import java.nio.file.Files;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static org.apache.parquet.hadoop.ParquetFileWriter.Mode.OVERWRITE;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * End‑to‑end PME flow: Parquet writer → Hive connector → Trino query.
+ */
+public class TestHiveParquetEncryption
+        extends AbstractTestQueryFramework
+{
+    private static final byte[] FOOTER_KEY = "footKeyIs16Byte?".getBytes(UTF_8);
+    private static final byte[] COLUMN_KEY = "colKeyIs16ByteA?".getBytes(UTF_8);
+    private static final ColumnPath AGE_PATH = ColumnPath.fromDotString("age");
+
+    /**
+     * kept so we can reference the warehouse path later
+     */
+    private java.nio.file.Path warehouseDir;
+
+    @Override
+    protected DistributedQueryRunner createQueryRunner()
+            throws Exception
+    {
+        warehouseDir = Files.createTempDirectory("pme_hive");
+
+        Map<String, String> properties = ImmutableMap.<String, String>builder()
+                .put("hive.metastore", "file")
+                .put("hive.metastore.catalog.dir", warehouseDir.toUri().toString())
+                // disable env retriever; we inject keys via custom module
+                .put("pme.environment-key-retriever.enabled", "false")
+                .buildOrThrow();
+
+        return HiveQueryRunner.builder()
+                .setHiveProperties(properties)
+                .setHiveModule(new TestingParquetEncryptionModule(FOOTER_KEY, COLUMN_KEY))
+                .build();
+    }
+
+    @Test
+    public void testEncryptedParquetRead()
+            throws Exception
+    {
+        // 1) write encrypted file inside warehouse
+        java.nio.file.Path dataDir = Files.createDirectories(warehouseDir.resolve("pme_data"));
+        java.nio.file.Path parquetFile = dataDir.resolve("data.parquet");
+        writeEncryptedFile(parquetFile);
+
+        // 2) create external table
+        String location = Location.of(String.valueOf(dataDir.toUri())).toString();
+        assertUpdate("""
+                CREATE TABLE enc_age(age INT)
+                WITH (external_location = '%s', format = 'PARQUET')
+                """.formatted(location));
+
+        // 3) verify results
```

Comment:
we also need to verify that an encrypted file cannot be read without encryption keys. e.g. a test where the footer key or the column key are not provided to the reader, resulting in an exception.